File: java/socketd/src/main/java/org/noear/socketd/transport/core/fragment/FragmentHandlerBase.java
package org.noear.socketd.transport.core.fragment;

import org.noear.socketd.transport.core.*;
import org.noear.socketd.transport.core.entity.EntityDefault;
import org.noear.socketd.transport.stream.StreamInternal;
import org.noear.socketd.utils.IoConsumer;

import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.MappedByteBuffer;

/**
 * 数据分片处理基类
 *
 * @author noear
 * @since 2.0
 */
public abstract class FragmentHandlerBase implements FragmentHandler {
    /**
     * 获取下个分片
     *
     * @param channel 通道
     * @param message 总包消息
     */
    @Override
    public void spliFragment(Channel channel, StreamInternal stream, MessageInternal message, IoConsumer<Entity> consumer) throws IOException {
        if (message.dataSize() > channel.getConfig().getFragmentSize()
                || message.data() instanceof MappedByteBuffer) {
            // MappedByteBuffer 的数据，也需要提前转出来
            int fragmentTotal = message.dataSize() / channel.getConfig().getFragmentSize();
            if (message.dataSize() % channel.getConfig().getFragmentSize() > 0) {
                fragmentTotal++;
            }

            int fragmentIndex = 0;
            while (true) {
                //生产分片
                fragmentIndex++;
                ByteBuffer dataBuffer = readFragmentData(message.data(), channel.getConfig().getFragmentSize());
                if (dataBuffer == null || dataBuffer.limit() == 0) {
                    return;
                }

                EntityDefault fragmentEntity = new EntityDefault().dataSet(dataBuffer);
                if (fragmentIndex == 1) {
                    fragmentEntity.metaMapPut(message.metaMap());
                }

                if (fragmentTotal > 1) {
                    //数量大于1才算真正的分版
                    fragmentEntity.metaPut(EntityMetas.META_DATA_FRAGMENT_IDX, String.valueOf(fragmentIndex));
                    fragmentEntity.metaPut(EntityMetas.META_DATA_FRAGMENT_TOTAL, String.valueOf(fragmentTotal));
                }

                consumer.accept(fragmentEntity);
                if (stream != null) {
                    stream.onProgress(true, fragmentIndex, fragmentTotal);
                }
            }
        } else {
            consumer.accept(message);
            if (stream != null) {
                stream.onProgress(true, 1, 1);
            }
        }
    }

    /**
     * 聚合所有分片
     *
     * @param channel       通道
     * @param fragmentIndex 分片索引（传过来信息，不一定有顺序）
     * @param message       分片消息
     */
    @Override
    public Frame aggrFragment(Channel channel, int fragmentIndex, MessageInternal message) throws IOException {
        FragmentAggregator aggregator = channel.getAttachment(message.sid());
        if (aggregator == null) {
            aggregator = createFragmentAggregator(message);
            channel.putAttachment(aggregator.getSid(), aggregator);
        }

        aggregator.add(fragmentIndex, message);

        if (aggregator.getDataLength() > aggregator.getDataStreamSize()) {
            //长度不够，等下一个分片包
            return null;
        } else {
            //重置为聚合帖
            channel.putAttachment(message.sid(), null);
            return aggregator.get();
        }
    }

    /**
     * 创建分片聚合器
     */
    protected abstract FragmentAggregator createFragmentAggregator(MessageInternal message) throws IOException;

    /**
     * 读取分版数据
     */
    protected ByteBuffer readFragmentData(ByteBuffer ins, int maxSize) {
        int size = 0;
        if (ins.remaining() > maxSize) {
            size = maxSize;
        } else {
            size = ins.remaining();
        }

        if (size == 0) {
            return null;
        }

        byte[] bytes = new byte[size];
        ins.get(bytes);

        return ByteBuffer.wrap(bytes);
    }
}

File: python/socketd/socketd/transport/core/fragment/FragmentHandlerBase.py
from abc import ABC, abstractmethod
from io import BytesIO
from typing import Optional, Callable

from socketd.transport.core.EntityMetas import EntityMetas
from socketd.transport.core.Frame import Frame
from socketd.transport.core.Channel import Channel
from socketd.transport.core.Message import Message, MessageInternal
from socketd.transport.core.entity.EntityDefault import EntityDefault
from socketd.transport.core.FragmentAggregator import FragmentAggregator
from socketd.transport.core.fragment.FragmentAggregatorDefault import FragmentAggregatorDefault
from socketd.transport.core.FragmentHandler import FragmentHandler
from socketd.transport.stream.Stream import StreamInternal


class FragmentHandlerBase(FragmentHandler, ABC):

    async def split_fragment(self, channel: Channel, stream: StreamInternal, message: MessageInternal, consumer: Callable):
        if message.data_size() > channel.get_config().get_fragment_size():
            fragment_total = message.data_size()  # channel.get_config().fragment_size
            if message.data_size() % channel.get_config().get_fragment_size() > 0:
                fragment_total += 1

            fragment_index = 0
            while True:
                # Generate fragment
                fragment_index += 1
                data_buffer = FragmentHandlerBase.__read_fragment_data(message.data(), message.data_size(),
                                                                       channel.get_config().get_fragment_size())
                if data_buffer is None or len(data_buffer.getbuffer()) == 0:
                    return

                fragment_entity: EntityDefault = EntityDefault().data_set(data_buffer)
                if fragment_index == 1:
                    fragment_entity.meta_map_put(message.meta_map())
                fragment_entity.meta_put(EntityMetas.META_DATA_FRAGMENT_IDX, str(fragment_index))
                fragment_entity.meta_put(EntityMetas.META_DATA_FRAGMENT_TOTAL, str(fragment_total))

                await consumer(fragment_entity)
                if stream is not None:
                    stream.on_progress(True, fragment_index, fragment_total)
                data_buffer.close()

        else:
            await consumer(message)
            if stream is not None:
                stream.on_progress(True, 1, 1)

    def aggr_fragment(self, channel: Channel, index: int, message: MessageInternal) -> Optional[Frame]:
        aggregator = channel.get_attachment(message.sid())
        if aggregator is None:
            aggregator = FragmentAggregatorDefault(message)
            channel.put_attachment(message.sid(), aggregator)

        aggregator.add(index, message)

        if aggregator.get_data_length() > aggregator.get_data_stream_size():
            return None  # Length is not enough, wait for the next fragment package
        else:
            channel.put_attachment(message.sid(), None)
            return aggregator.get()  # Reset as a merged frame

    @abstractmethod
    def create_fragment_aggregator(self, message: MessageInternal) -> FragmentAggregator:
        """创建分片聚合器"""
        ...

    @staticmethod
    def __read_fragment_data(ins: BytesIO, ins_size: int, max_size: int) -> BytesIO:
        size = min(ins_size - ins.tell(), max_size)
        buf = BytesIO()
        buf.write(ins.read(size))
        return buf
